packages/xstate-solid/src/stateUtils.ts

```diff
@@ -0,0 +1,83 @@
+import type { SetStoreFunction } from 'solid-js/store';
+import { reconcile } from 'solid-js/store';
+import type { ActorRef, Interpreter, InterpreterFrom } from 'xstate';
+import { State } from 'xstate';
+import { isStateLike } from 'xstate/lib/utils';
+
+function isService(value: any): value is Interpreter<any> {
+  return 'state' in value && 'machine' in value;
+}
+
+/**
+ * Reconcile the state of the machine with the current state of the store.
+ * Handles primitive values, arrays, and objects.
+ * Provides granular reactivity for the state of the machine in SolidJS.
+ * @param nextState The next state value to update current store with
+ * @param setState A Solid store setter
+ */
+export const updateState = <NextState extends object | unknown>(
+  nextState: NextState,
+  setState: SetStoreFunction<NextState>
+): void => {
+  // Only reconcile each property if nextState is a State class
+  if (nextState instanceof State) {
```

Comment:
To deeply update our reactive State (because it's a class), we need to reconcile each property rather than the root class instance. There are 5 tests that cover this